sky/serve/core.py

```diff
@@ -91,6 +96,131 @@ def _validate_service_task(task: 'sky.Task') -> None:
+    """Helper function of sync_down_logs.
+        - prepare and download load balancer / controller log file.
+    """
+    logger.info('Starting the process to prepare '
+                f'and download {service_component.name.lower()} logs...')
+    runner = controller_handle.get_command_runners()
+
+    if service_component == serve_utils.ServiceComponent.CONTROLLER:
+        log_file_name = serve_utils.generate_remote_controller_log_file_name(
+            service_name)
+        log_file_constant = serve_constants.CONTROLLER_LOG_FILE_NAME
+    elif service_component == serve_utils.ServiceComponent.LOAD_BALANCER:
+        log_file_name = serve_utils.generate_remote_load_balancer_log_file_name(
+            service_name)
+        log_file_constant = serve_constants.LOAD_BALANCER_LOG_FILE_NAME
+    else:
+        with ux_utils.print_exception_no_traceback():
+            raise ValueError(
+                f'Unsupported service component: {service_component}')
+
+    log_files_to_sync.append((log_file_name, log_file_constant))
+
+    if not sync_down_all_components:
+        single_file_synced = True
+        target_directory = os.path.join(target_directory, log_file_constant)
+
+    return runner, target_directory, single_file_synced
+
+
+def _sync_down_replica_logs(service_name, run_timestamp, replica_id,
+                            controller_handle, sky_logs_directory) -> None:
+    """Helper function of sync_down_logs.
+        - prepare and download replica log file.
+    """
+    logger.info('Starting the process to prepare and download replica logs...')
+    prepare_code = serve_utils.ServeCodeGen.prepare_replica_logs_for_download(
+        service_name, run_timestamp, replica_id)
+
+    backend = backend_utils.get_backend_from_handle(controller_handle)
+    assert isinstance(backend, backends.CloudVmRayBackend)
+    assert isinstance(controller_handle, backends.CloudVmRayResourceHandle)
+
+    # Prepare replica logs for download.
+    logger.info('Preparing replica logs for download on controller...')
+    prepare_returncode = backend.run_on_head(controller_handle,
+                                             prepare_code,
+                                             require_outputs=False,
+                                             stream_logs=False)
+    logger.debug('Prepare command return code: %s', prepare_returncode)
+
+    _run_command_with_error_handling(
+        prepare_returncode, prepare_code,
+        'Failed to prepare replica logs to sync down. '
+        'Sky serve controller may be older and not support sync '
+        'down for replica logs.')
+
+    # Define remote directory for logs and download them.
+    runner = controller_handle.get_command_runners()
+    remote_service_dir_name = serve_utils.generate_remote_service_dir_name(
+        service_name)
+    _sync_log_file(runner,
+                   remote_service_dir_name,
+                   sky_logs_directory,
+                   run_timestamp=run_timestamp)
```

Comment:
Is it possible that we prepare all the logs (the replica log, controller log & lb log), construct all targets to download, and loop them once, and then cleanup?